File: go/srp.go
// SRP (Single Responsibility Principle)
// Separate the code into simple, well-defined, well-intentioned tasks and give clear names.

package main

import (
	"encoding/json"
	"fmt"
	"strings"
)

// Before
type ReportGeneratorBefore struct{}

func (r *ReportGeneratorBefore) GenerateReport(data []string) {
	// Process the data
	processedData := make([]string, len(data))
	for i, item := range data {
		processedData[i] = strings.ToUpper(item)
	}

	// Convert data to JSON
	jsonData, _ := json.Marshal(processedData)

	// Save the report to a file (simulated)
	fmt.Printf("Saving to file: %s\n", string(jsonData))

	// Send the report via email (simulated)
	fmt.Printf("Email sent with data: %s\n", string(jsonData))
}

// After
type ReportGenerator struct{}

func (r *ReportGenerator) GenerateReport(data []string) {
	processedData := r.processData(data)
	jsonData := r.convertToJSON(processedData)
	r.saveReport(jsonData)
	r.sendReport(jsonData)
}

func (r *ReportGenerator) processData(data []string) []string {
	processedData := make([]string, len(data))
	for i, item := range data {
		processedData[i] = strings.ToUpper(item)
	}
	return processedData
}

func (r *ReportGenerator) convertToJSON(data []string) string {
	jsonData, _ := json.Marshal(data)
	return string(jsonData)
}

func (r *ReportGenerator) saveReport(jsonData string) {
	fmt.Printf("Saving to file: %s\n", jsonData)
}

func (r *ReportGenerator) sendReport(jsonData string) {
	fmt.Printf("Email sent with data: %s\n", jsonData)
}

func main() {
	data := []string{"a", "b", "c", "d"}
	
	reportBefore := &ReportGeneratorBefore{}
	reportBefore.GenerateReport(data)

	fmt.Println()

	report := &ReportGenerator{}
	report.GenerateReport(data)
}

File: cpp/srp.cpp
// SRP (Single Responsibility Principle)
// Separate the code into simple, well-defined, well-intentioned tasks and give clear names.

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

// Before
class ReportGeneratorBefore {
public:
    void generateReport(const std::vector<std::string>& data) {
        // Process the data
        std::vector<std::string> processedData;
        for (const auto& item : data) {
            std::string processed = item;
            std::transform(processed.begin(), processed.end(), processed.begin(), ::toupper);
            processedData.push_back(processed);
        }

        // Convert data to JSON (simplified)
        std::cout << "Saving to file: [";
        for (size_t i = 0; i < processedData.size(); ++i) {
            std::cout << "\"" << processedData[i] << "\"";
            if (i < processedData.size() - 1) std::cout << ", ";
        }
        std::cout << "]" << std::endl;

        // Send the report via email (simulated)
        std::cout << "Email sent with data: [";
        for (size_t i = 0; i < processedData.size(); ++i) {
            std::cout << "\"" << processedData[i] << "\"";
            if (i < processedData.size() - 1) std::cout << ", ";
        }
        std::cout << "]" << std::endl;
    }
};

// After
class ReportGenerator {
public:
    void generateReport(const std::vector<std::string>& data) {
        auto processedData = processData(data);
        auto jsonData = convertToJson(processedData);
        saveReport(jsonData);
        sendReport(jsonData);
    }

private:
    std::vector<std::string> processData(const std::vector<std::string>& data) {
        std::vector<std::string> processedData;
        for (const auto& item : data) {
            std::string processed = item;
            std::transform(processed.begin(), processed.end(), processed.begin(), ::toupper);
            processedData.push_back(processed);
        }
        return processedData;
    }

    std::string convertToJson(const std::vector<std::string>& data) {
        std::string json = "[";
        for (size_t i = 0; i < data.size(); ++i) {
            json += "\"" + data[i] + "\"";
            if (i < data.size() - 1) json += ", ";
        }
        json += "]";
        return json;
    }

    void saveReport(const std::string& jsonData) {
        std::cout << "Saving to file: " << jsonData << std::endl;
    }

    void sendReport(const std::string& jsonData) {
        std::cout << "Email sent with data: " << jsonData << std::endl;
    }
};

int main() {
    std::vector<std::string> data = {"a", "b", "c", "d"};
    
    ReportGeneratorBefore reportBefore;
    reportBefore.generateReport(data);

    std::cout << std::endl;

    ReportGenerator report;
    report.generateReport(data);

    return 0;
}
